src/pmhn/_trees/_backend.py

```diff
@@ -57,26 +59,131 @@ def gradient_and_loglikelihood(
+                        set(
+                            [
+                                tree_mutation.name  # type: ignore
+                                for tree_mutation in tree_mutations
+                            ]
+                        )
+                    )
+                )
+                for mutation in exit_mutations:
+                    lamb = 0
+                    exit_subclone = [
+                        anc.name  # type: ignore
+                        for anc in node.path
+                        if anc.parent is not None
+                    ] + [mutation]
+                    for j in exit_subclone:
+                        lamb += theta[mutation - 1][j - 1]
+                    lamb = np.exp(lamb)
+                    lamb_sum -= lamb
+                for child in node.children:
+                    next_nodes.append(child)
+            current_nodes = next_nodes
+        return lamb_sum
+
+    def off_diag_entry(self, tree1: Node, tree2: Node, theta: np.ndarray) -> float:
+        """
+        Calculates an off-diagonal entry of the V matrix.
+
+        Args:
+            tree1: the first tree
+            tree2: the second tree
+            theta: real-valued (i.e., log-theta) matrix,
+              shape (n_mutations, n_mutations)
+
+        Returns:
+            the off-diagonal entry of the V matrix corresponding to tree1 and tree2
+        """
+        exit_node = bfs_compare(tree1, tree2)
+        lamb = 0
+        if exit_node is None:
+            return lamb
+        else:
+            for j in [
+                node.name  # type: ignore
+                for node in exit_node.path
+                if node.parent is not None
+            ]:
+                lamb += theta[exit_node.name - 1][j - 1]
+            lamb = np.exp(lamb)
+            return float(lamb)
+
     def loglikelihood(
-        self,
-        tree: Tree,
-        theta: np.ndarray,
+        self, tree: Node, theta: np.ndarray, sampling_rate: float
     ) -> float:
-        """Calculates loglikelihood `log P(tree | theta)`.
+        """
+        Calculates loglikelihood `log P(tree | theta)`.
 
         Args:
             tree: a tree
             theta: real-valued (i.e., log-theta) matrix,
               shape (n_mutations, n_mutations)
-
+            sampling_rate: a scalar of type float
         Returns:
             loglikelihood of the tree
         """
         # TODO(Pawel): this is part of https://github.com/cbg-ethz/pMHN/issues/15
         #   It can be implemented in any way.
-        raise NotImplementedError
+        V = self.create_V_Mat(tree=tree, theta=theta, sampling_rate=sampling_rate)
+        V_size = V.shape[0]
+        b = np.zeros(V_size)
+        b[0] = 1
+        V_transposed = V.transpose()
+        V_csr = csr_matrix(V_transposed)
+        x = spsolve_triangular(V_csr, b, lower=True)
+
+        return np.log(x[V_size - 1] + 1e-10) + np.log(sampling_rate)
```

Comment:
The `1e-10` jitter is a [magic constant](https://en.wikipedia.org/wiki/Magic_number_(programming)).

I'd suggest to set it in the `__init__(self, jitter: float = 1e-10)` function, so that (a) it's not magic constant anymore and (b) we can later check whether modifying it makes any changes.